### ISSUE TITLE
Unhelpful / misleading error message and help message when using a reference to a vector

### ISSUE THREAD

**Issue by njor** [rust-lang/rust#83924]
Hello,
There is a case where the Rust compiler error message is misleading. I tried this code :

```rust
fn main() {
    let mut values = vec![10, 11, 12];
    let v = &mut values;

    let mut max = 0;
    
    for n in v {
        max = std::cmp::max(max, *n);
    }

    println!("max is {}", max);
    println!("Converting to percentages of maximum value...");
    for n in v {
        *n = 100 * (*n) / max;
    }
    println!("values: {:#?}", values);
}
```

…which gives the following error message, and misleading « help » message :
```
error[E0382]: use of moved value: `v`
   --> src/main.rs:13:14
    |
3   |     let v = &mut values;
    |         - move occurs because `v` has type `&mut Vec<i32>`, which does not implement the `Copy` trait
...
7   |     for n in v {
    |              -
    |              |
    |              `v` moved due to this implicit call to `.into_iter()`
    |              help: consider borrowing to avoid moving into the for loop: `&v`
...
13  |     for n in v {
    |              ^ value used here after move
    |
```
Especially, the `help: consider borrowing to avoid moving into the for loop: '&v'` is the misleading part, does not help solve the problem at all, and sent me an a wrong path.
As I am new to Rust, this got me scratching my head for some time until someone explained it to me ([here](https://users.rust-lang.org/t/use-of-moved-value-when-using-a-reference-to-a-vector/57873/2)).
(And since the Rust compiler is most of the time very helpful, I did not even consider that its suggestion could be wrong haha. Hm.)
Maybe a short message explaining how to reborrow the reference in this kind of situation would be more helpful and less misleading ?
Thank you.

### Meta

`rustc --version --verbose`:
```
rustc 1.48.0 (7eac88abb 2020-11-16)
binary: rustc
commit-hash: 7eac88abb2e57e752f3302f02be5f3ce3d7adfb4
commit-date: 2020-11-16
host: x86_64-unknown-linux-gnu
release: 1.48.0
LLVM version: 11.0
```

**Comment by eggyal** [comment]
I'm actually surprised that you got E0277 when doing `for n in &v` since I'd have expected that to resolve (via deref coercion) to the impl of `IntoIterator` for `&Vec`.

**Comment by eggyal** [comment]
> I'm actually surprised that you got E0277 when doing `for n in &v` since I'd have expected that to resolve (via deref coercion) to the impl of `IntoIterator` for `&Vec`.

Okay, this is due to the desugaring in `rustc_ast_lowering::expr::lower_expr_for`:

https://github.com/rust-lang/rust/blob/e1d49aaad4fde33469fdb786c29838a95a5d8a11/compiler/rustc_ast_lowering/src/expr.rs#L1710-L1728

Because `into_iter` is invoked as a trait function `IntoIterator::into_iter(<head>)` rather than a method, deref coercion does not occur.  It would be nicer if this could instead desugar to `<head>.into_iter()` but I guess there could then be issues correctly resolving to `IntoIterator`—but it shouldn't be impossible to ensure that happens correctly?
